feat(dhall-docs): Specification and implementation of the dhall-docs comments format (#1929)
dhall-lang/dhall-haskell: dhall-docs/README.md

# `dhall-docs`

> :construction: **This tool is on development phase yet. You can use it but you may expect breaking changes on each new release until the 1.0 release.**

Reviewer: I would personally not worry keeping all breaking changes before the 1.0 release.  Any tool that has enough users will require a slow stream of slightly breaking changes over time to keep up with their evolving needs.

The rule of thumb I always use is that the users decide when a given tool is out of beta, not the author.  This is one of the reasons I always make the first published version of my packages 1.0.0, in recognition of the fact that users might put my tool into production regardless of whether or not I warn them not to.

Author: I understand your suggestion, thanks!

I think I'll leave the warning about the current status of the tool, it may be a good notice to people though:

```suggestion
:construction: **This tool is on development phase yet**
```

> :construction: **This tool is on development phase yet. You can use it but you should expect a lot of breaking changes on each new release until the final release.**

Reviewer 2: ```suggestion
:construction: **This tool is on development phase yet. You can use it but you should expect a lot of breaking changes on each new release until the 1.0 release.**
```

?

You can see examples of writing documentation on the [tests](./tasty/data/package)
folder.

### Block comments

Normal block comments in Dhall starts with `{-` and ends with `-}`. For example:

```dhall
{- foo
     bar
       baz
-}
```

A `dhall-docs` block comment needs to start with `{-|<NEW_LINE>`.
That is, the actual documentation starts on the following line. The closing brace
doesn't need any special treatment. This let you separate your own internal block
comments from the ones that you want `dhall-docs` to output in the final markup.

[CommonMark] supports some features that are sensitive to indentation (e.g.
[indented-code-blocks]). On a `dhall-docs` block comment, the indentation is
determined by the column of the `{` character from the opening brace. For
example, on the following block comment:

```dhall
{-|
foo
  bar
    baz
-}
```

... and the following:

```dhall
              {-|
              foo
                bar
                  baz
                        -}
```

... will extract the contents of the comment, stripping the `{` leading whitespace
from all other lines, returning the following text:

```
foo
  bar
    baz
```

> 

Reviewer 2: I think "indentation" should be defined more clearly. Can it be a mix of tabs and spaces?

The indentation level of a line is the number of whitespace characters (i.e. `' '`) from

> the first column until the column of the `{`. The rest of the whitespace that

Reviewer 2: If tabs and space character are interchangeable, that means that e.g.

```
	{-|
 foo
 -}
```

is well-formed, right?

I don't know whether this could cause problems, but I think it might be worth spelling out.

Reviewer: I think a lot of these issues would be fixed by using the exact same rules as multi-line strings in Dhall.  Then there's no confusion

Reviewer 2: I think Gabriel has a point here. If we can specify how a doc comment corresponds to a multi-line string, we only need to add that the corresponding string is supposed to be the input for Markdown parser, and we're done with a good part of this spec.

I guess the tricky bit is that the multi-line spec isn't exactly super easy to understand itself.

How familiar are you with it at this stage, @german1608?

Author: @sjakobi: yes, that example is well formed.

@Gabriel439: I thought about sing something similar to multiline strings for this purpose, but I was thinking about something more strict for the comments and ended up using {-| as base of indentation.

Another idea that I had was to use the first line column as a base of indentation, that is more similar to multiline strings. But it will allow users to write things such as:

```
{-|
                  Foo
                      BAR
-}
```

That doesn't look to good IMO

is not stripped will be preserved for further manipulations (i.e. converting
[CommonMark] to HTML).

A block comment that has a line with less indentation than the `{` character
column one is not a valid `dhall-docs` comment. The following:

```dhall
    {-|
    foo
  bar
        baz -}

> ```

Reviewer: What about blank lines without leading whitespace?  Are those lines ignored for the purpose of computing indentation?

The reason I ask is that multi-line strings will ignore such lines, and it might be worth documenting whether documentation comments do the same.  Ideally, the more the documentation format should resemble how multi-line strings work in Dhall

Author: Good catch. I believe we should ignore them as well.

```dhall
{-| foo
    bar
    -}
```

is invalid.

See that the closing brace (`-}`) can be in a new line or in the last one:
`dhall-docs` will, in any case, insert a newline character (`\n`) in the last

> line.

Author: Now reading it I see that I didn't reflect my thoughts about it.

If the `-}` is on a new line e.g.

```
{-|
a
-}
```

or like this:

```
{-|
a-}
```

the text will be extracted like this `"a\n"`. My thought is that it would be good for the user that he knows he can place the `-}` at any place and the result will be the same.

is invalid.

See that the closing brace (`-}`) can be in a new line or in the last one:
`dhall-docs` will, in any case, insert a newline character (`\n`) in the last
line. The following example:

```dhall
{-|
foo-}
```

and

```dhall
{-|
foo
-}

> ```

Reviewer: Perhaps not insert an implicit newline for the former case.  The only reason I suggest this is so that the documentation format more closely follows how multi-line strings work in Dhall.

Author: I'll remove that fragment of the documentation. I think it confuses the user.

> will both extract `"foo\n"`.

Reviewer 2: Is the trailing `\n` relevant for the generated documentation? Otherwise this seems unnecessarily detailed. Maybe just say that the two examples are equivalent with regard to the generated documentation.

Author: Mmm now that you say it I'm unsure. I'll remove it and add it if it's necessary.

> will both extract `"foo\n"`.

Reviewer 2: Also, is there a difference between

```
{-|
foo
-}
```

and

```
{-|
foo

-}
```

?

Author: I'd say they are different. Or at least the tool should not strip the newlines. My reasoning is that Markdown is a little sensitive to whitespaces, and after thinking a little it would be better to do the less whitespace stripping on the text. The only exception would be the prefix whitespace for indentation, that is mostly discussed on the proposal.

in
{
  {-|
  foo
      bar
          baz
  -}
  myAwesomeFunction
}
```

> If a block comment doesn't starts with `{-|` it will be ignored by `dhall-docs`.

Reviewer 2: (Actually my suggestion doesn't seem so much clearer… feel free to ignore)

> If a block comment doesn't starts with `{-|` it will be ignored by `dhall-docs`.

Author: I'll ignore your suggestion but I found a typo on mine: `starts` -> `start`.

> 

Reviewer 3: It may be useful to include some more common usage such as:

* Bold, italic
* Internal (anchor) link
* Hypertext link
* Table (is the CommonMark parser support them?)

Author: Since the markup language is based on Markdown (specifically, CommonMark), those features will be supported out-of-the-box. We could write a sample of that on that example, though.

The only thing that we need to discuss is how to link to other files on the package, but we could do that in a separate issue.

Reviewer 3: If I understand correctly, CommonMark does not yet have a syntax for tables, accoding to this [thread](https://talk.commonmark.org/t/tables-in-pure-markdown/81), and for example when using Pandoc you need to activate the `githubMarkdownExtensions` when parsing CommonMark.

The samples doesn't have to be exhaustive, for example the `foo bar baz` sample could be replaced by:

<pre>
* a list item in *bold*
  - a sub item in _italic_
  - a [link](http://example.com)
</pre>

Author: Oh sorry!

mmark, out markdown parser, doesn't support them AFAIK, but we could create a custom extension to add that feature if it's needed.

the rules on the specification, it will be ignored _and_ a warning will be
logged by `dhall-docs`.

### Line comments

Normal Dhall single-line comments starts with `--` and ends at a newline. For
example:

```dhall
-- single line comment
0
```

A `dhall-docs` valid single-line comment starts with `--|` and a single whitespace
character. For example:

```dhall
--| foo
```

You can span your documentation in several line comments by writing `--` and two
(2) whitespaces:

```dhall
--| foo
--  bar
--      baz
```

The `--` prefix defines the indentation for any line. That is, on the above
example and the following

```dhall
--| foo
  --  bar
 --      baz
```

... `dhall-docs` will extract the following contents:

```
foo
bar
    baz

> ```

Reviewer 2: These rules seem quite different than those for block comments, but after thinking about them for a bit, I haven't come up with anything better. :+1:

Author: yeah, they are quite different. The main difference is indentation treatment on both of them, but couldn't think any better that is easy to write and to implement than this.

--| foo
--  bar
--      baz
```

The `--` prefix defines the indentation for any line. That is, on the above
example and the following

```dhall
--| foo
  --  bar

>  --      baz

Reviewer: I think the documentation generator is being too lenient with respect to indentation.  I would prefer that the documentation generator enforces the recommended comment layout and rejects everything else, otherwise it will be easy for people to inadvertently commit the wrong docs.

Also, in my experience, it's easier to start off strict and then make things more lenient later on in response to user feedback than to do things the other way around (start off too lenient and then make things more strict)

Author: I agree. I'll update the convention to enforce the alignment of `--`

```dhall
foo

bar
```

The empty `--` line doesn't need the two (2) whitespaces.

If you left an empty line on the source code between two lines, it will be ignored.
On this example:

```dhall
--| foo

--  bar

```

`dhall-docs` will extract:

```
foo
bar

> ```

Reviewer 2: I'd rather suggest treating `bar` as a separate non-documentation comment. Otherwise I think we'll find "internal" comments appended to the documentation.

```dhall
--| # My header
--
--  * 1
--  * 2

let a --| foo
        --  bar
        = True

in { a }
```

Although `dhall-docs` doesn't enfoce an alignment between the set of consecutive
single-line comments, we encourage you to write them aligned. The above example
written with this recommendation looks like this:

```dhall
--| # My header
--
--  * 1
--  * 2

let a --| foo
      --  bar
        = True

in { a }

```

To end, if the first line of a single-line comments doesn't start with `--|` it
will be ignored by `dhall-docs`. On the other hand if the first line starts with
`--|` but the set of lines doesn't follow the rules on this specification, it
will be ignored _and_ a warning will be logged to the console.

### Supported annotated elements

A limitation of the `dhall` parser is that it doesn't preserve _all_ the whitespaces
of a Dhall source file, but preserves the enough to write useful documentation.
`dhall-docs` supports the following comments:

* The header of a file:

    ```dhall
    {-| this is the header
    -}
    let a = 1
    in  a
    ```

* Before each key-value pair of a record type and literal:

    ```dhall
    {
      --| here
      a : T,
    }

    {

>       -- | and here

Reviewer 2: That's illegal!!! ;)

```suggestion
      --| and here
```

Author: The Dhall standard says it's valid:

https://github.com/dhall-lang/dhall-lang/blob/5fde8ef1bead6fb4e999d3c1ffe7044cd019d63a/standard/dhall.abnf#L167-L169

      pun,

      -- here too
      b = True
    }

>     ```

Reviewer 2: I think we should maybe wait with this until we've worked out what `dhall-docs` will do with these annotations.

For more information about the tool, check the `--help` flag.

## Documenting your packages

Documenting your package is essentially writing comments on your source code,
although there are some format rules that these needs to obey in order to work
properly with `dhall-docs`. These rules aid `dhall-docs` in extracting the text
on the documentation that will be passed to our Markdown preprocessor to finally
render your documentation in HTML.

You can see examples of writing documentation on the [tests](./tasty/data/package)
folder.

### Documentation markup language

The markup language of the documentation is [CommonMark], which is a strict
Markdown flavour. `dhall-docs` uses [`mmark`], a package that parses
[CommonMark] and transforms it to HTML. This package follows (almost)
that specification.

### Block comments

Normal block comments in Dhall starts with `{-` and ends with `-}`. For example:

```dhall
{- foo
     bar
       baz
-}
```

A `dhall-docs` block comment needs to start with `"{-|\n"` i.e. a `{-|` and a

> newline. This means that the actual documentation _must_ start on the following line.

Reviewer 2: In analogy to multi-line strings, we should also support `\r\n` line-endings:

https://github.com/dhall-lang/dhall-lang/blob/f717c748b3751e5a1b22da8af3db31030e7dcd0d/standard/dhall.abnf#L348

[CommonMark] supports some features that are sensitive to indentation (e.g.
[indented-code-blocks]). Indentation on `dhall-docs` block comments is really
similar to [Dhall multi-line strings]: the indentation is determined by the
longest common whitespace prefix between the lines on the block comment. The
main difference is that, unlike the closing pair of double single quotes on multi-line

> strings (`''`), the closing brace (`-}`) doesn't alter indentation.

Author: From my POV it would be annoying if me, a user, should care about the position of the closing brace `-}`

> strings (`''`), the closing brace (`-}`) doesn't alter indentation.

Reviewer 2: Hmm, I'm not convinced that this is important enough that it would be worth breaking the analogy for. There's also at least one case where the significance of the position of the `-}` would be useful:

```
{-
    indented
    code
    block
-}
```

> strings (`''`), the closing brace (`-}`) doesn't alter indentation.

Author: Good point :) then a `dhall-docs` block comment will be basically a multi-string :+1:

on the extracted text. On the following example:

```dhall
  {-|
    foo

    bar
    -}
```

`dhall-docs` will extract the following text:

    foo

    bar

Note that a line that only contains trailing whitespace will be taken into
account when analyzing the common indentation. The line between `foo` and `bar`
on the following example has leading whitespace (two (2) whitespaces):

```dhall
  {-|
    foo
  
    bar -}
```

meaning that the extracted text will be:

```
  foo

  bar
```

If there is no newline between the opening brace of a `dhall-docs` block comment
(i.e. `{-|`), that comment will be considered invalid: `dhall-docs` will not
render it on the documentation _and_ a warning will be logged on the console.
The following:

```dhall
  {-| foo
      bar -}
```

although it's a valid Dhall block comment, it's invalid for `dhall-docs`.

If you write a comment (either block or single-line comments) after a valid
`dhall-docs` comment, it will be ignored. Wheter they were valid `dhall-docs`
comments or not. On the two following examples:

```dhall
{-|
  foo -}
-- bar
```

```dhall
{-|
  foo
-}

{-|
-}
```

> the extracted text will only contain `foo`.

Author: Ok, but I'd say that we should still keep ignoring non-doc comments after a doc comment.

What about something like this:

```
{-|
foo -}
-- bar
{-|
baz -}
```
should be an error?

> the extracted text will only contain `foo`.

Reviewer 2: Yeah, I think that should be an error too. It's just not clear which doc comment the user intended to be used by `dhall-docs`.

AFAIK haddock concatenates the doc comments in such cases, but I think that's error prone and not really useful in practice.

> the extracted text will only contain `foo`.

Author: > Yeah, I think that should be an error too. It's just not clear which doc comment the user intended to be used by `dhall-docs`.

Ok, so the rule is: after a `dhall-docs` comment, there can't be any other `dhall-docs` comment following it.

character. For example:

```dhall
--| foo
```

You can span your documentation in several single-line comments by writing `--` and two
(2) whitespaces:

```dhall
--| foo
--  bar
--      baz
```

The `--` prefix defines the indentation for any line. That is, on the above
example `dhall-docs` will extract the following contents:

```
foo
bar
    baz
```

If you would like to force an empty line on the documentation output, place an
empty single line comment in that place. On this example:

```dhall
--| foo
--
--  bar
```

`dhall-docs` will extract:

```dhall
foo

bar
```

The empty `--` line doesn't need the two (2) whitespaces.

If between two single-line comments there is one (1) or more empty lines, the
comments after those will be ignored. On this example:

```dhall
--| foo
--  bar

--  baz
```

will only capture this text:

```
foo
bar
```

Also note that in a set of consecutive single line comments where the `|` is
not on the first one, the single-line comments before that one will be
ignored on the rendered documentation. On this example:

```dhall
--  foo
--  bar
--| baz
--  qux
```

the following text will be extracted:

```dhall
baz
qux
```

> The set of `--` lines needs to be aligned, meaning that a set of lines like this:

Reviewer 2: "aligned" is slightly vague. Is this aligned?

```
	--| foo
 -- bar
```

What does it mean when before the `--|` there are [fullwidth characters](https://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms)?

For simplicity, I think we can demand that each `--` is preceded by the same number of characters (if that's well-defined?!). Or we could give up the alignment requirement again…

Author: I agree with that definition of alignment. I'd say that will be enough for almost all cases.

I don't have a problem with no enforcing the alignment, but following @Gabriel439 comment ([here](https://github.com/dhall-lang/dhall-haskell/pull/1929#discussion_r456917946)) we should start with something more strict and make it more flexible depending on user requests.

```dhall
--| foo
  --  bar
 --  baz
```

is invalid: `dhall-docs` will ignore this on the generated documentation and
a warning will be logged on the console.

If the `|` is not after the first `--`, that comment will be ignored by `dhall-docs`
without logging anything to the console.

If you write a block comment after a set of `dhall-docs` comments, it will be
ignored, whether it was a valid `dhall-docs` comment or not. On the following
two examples:

```dhall
--| foo
--  bar
{- qux -}
```

```dhall
--| foo
--  bar
{-|
 qux -}
```

the extracted text will be the following:

```
foo
bar

> ```

Reviewer 2: As mentioned above, I think we should treat subsequent doc comments as errors.

Can we deduplicate these sections BTW?

Author: Sure! I'll add another section on the document

Author: I merged both sections in a single one and added some examples there. The section is `Mixing the two type of comments`

  bar
    baz
-}
```

```dhall
{-|
    foo
      bar
        baz
    -}
```

```dhall
{-|
    foo
      bar
        baz-}
```

... `dhall-docs` will extract the contents of the comment, stripping all the
common leading whitespace from all lines, returning the following text that will
be passed to our [CommonMark] parser:

```
foo
  bar
    baz
```

Similar to Dhall multi-line string literals, the `-}` position also affects the
final indentation, meaning that in the following sample:

```dhall
{-|
    foo
    bar
-}
```

The following text will be extracted:

```
    foo
    bar
```

Only spaces and tabs are taken into account in calculating the longest common
whitespace prefix. Other forms of whitespaces are not considered.

Empty lines are ignored when determining the indentation level but are preserved
on the extracted text. On the following example:

```dhall
{-|
foo

bar
-}
```

`dhall-docs` will extract the following text:

```
foo

bar
```

Note that a line that only contains trailing whitespace will be taken into
account when calculating the common indentation. The line between `foo` and `bar`
on the following example has leading whitespace (two (2) whitespaces):

```dhall

>   {-|
>     foo
>   
>     bar -}

Reviewer: It might help to use `␠` or `␣` to show the spaces:

```suggestion
␠␠{-|
␠␠␠␠foo
␠␠
␠␠␠␠bar -}
```

```dhall
--| foo 
--  bar

--  baz
```

`dhall-docs` will only capture this text:

```
foo
bar
```

In a set of subsequent single-line comments, all lines before the one that has
the `|` character will be ignored, meaning that if any line on the set doesn't
have the `|` marker, no text will be extracted for the generated documentation.

On this example:

```dhall
--  foo
--  bar
--| baz
--  qux
```

the following text will be extracted:

```dhall
baz
qux
```

and this set of single-line comments:

```dhall
--  foo
--  bar
--  baz
```

will be ignored by the tool.

The set of `--` lines needs to be aligned with the same number of preceding
characters, meaning that a set of lines like this:

```dhall
--| foo
  --  bar
 --  baz
```

is invalid: `dhall-docs` will ignore this on the generated documentation and
a warning will be logged on the console.

To end, if there are language tokens in-between a set of lines, the lines after
those tokens will be ignored. On this example:

```dhall
--| foo
,
--  bar
```

the extracted text will be the following:

    foo

Here is an example of using this type of comments in a Dhall file:

```dhall
--| # My header
--
--  * 1
--  * 2

> let a --| foo
>       --  bar
>         = True

Reviewer: ```suggestion
let --| foo
    --  bar
    a = True
```

I mainly suggest this since that's the commenting format we've been recommending so far for `let` bindings
